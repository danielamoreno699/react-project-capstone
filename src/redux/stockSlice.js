import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const apiURL = 'https://api.polygon.io/v3/reference/tickers?market=stocks&active=true&apiKey=Vp4PFtmRiJhA03_A8hWFal2787C86E4P';

// fetching stocks form home page
export const fetchData = createAsyncThunk('stocks/fetchData', async () => {
  const response = await fetch(apiURL);
  if (response.ok) {
    const data = await response.json();
    return data.results;
  }
  throw new Error('Failed to fetch data');
});

// fetching details of stock passing the parameter of stock
export const fetchStockInfo = createAsyncThunk(
    'stocks/fetchStockInfo',
    async (stock) => {
      const response = await fetch(`https://api.polygon.io/v3/reference/tickers/${stock}?apiKey=Vp4PFtmRiJhA03_A8hWFal2787C86E4P`);
      if (response.ok) {
        const data = await response.json();
        return data.results;
      }
      throw new Error('Failed to fetch stock information');
    },
  );

  const initialState = {
    status: 'idle',
    data: [],
    selectedStock: null,
    stockInfo: {},
  
  };
  
  export const stockSlice = createSlice({
    name: 'stocks',
    initialState,
    reducers: {
      selectStock: (state, action) => {
        const selectedStock = action.payload;
        return {
          ...state,
          selectedStock,
          stockInfo: {},
        };
      },
    },
    extraReducers: (builder) => {
        builder
          .addCase(fetchData.pending, (state) => {
            state.status = 'loading';
          })
          .addCase(fetchData.fulfilled, (state, action) => ({
            ...state,
            status: 'succeeded',
            data: action.payload,
          }))
          .addCase(fetchData.rejected, (state, action) => ({
            ...state,
            status: 'failed',
            error: action.error.message,
          }))
          .addCase(fetchStockInfo.fulfilled, (state, action) => ({
            ...state,
            status: 'succeeded',
            stockInfo: action.payload,
          }))
          .addCase(fetchStockInfo.rejected, (state, action) => ({
    
            ...state,
            status: 'failed',
            error: action.error.message,
          }));
      },
    });

    export const { selectStock } = stockSlice.actions;
export default stockSlice.reducer;