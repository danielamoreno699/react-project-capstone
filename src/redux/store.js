import { configureStore } from '@reduxjs/toolkit';

const store = configureStore({
  reducer: {
    stocks: stocksReducer,

  },
});

export default store;
