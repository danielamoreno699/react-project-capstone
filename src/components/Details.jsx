import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { BiArrowBack } from 'react-icons/bi';
import { useNavigate, useParams } from 'react-router-dom';
import { fetchStockInfo } from '../redux/stockSlice';
import '../styles/detail.css';

const Details = () => {
  const { stock } = useParams();
  const navigate = useNavigate();

  const onNavigateBack = () => {
    navigate(-1);
  };

  const dispatch = useDispatch();
  const { stockInfo } = useSelector((state) => state.stocks);

  useEffect(() => {
    dispatch(fetchStockInfo(stock));
  }, [dispatch, stock]);

  return (
    <>
      <div data-testid="details" className="detail-container">
        <div className="back-symbol">
          <BiArrowBack onClick={onNavigateBack} />
        </div>

        <div className="header-detail">

          <div className="txt-container">

            <div className="txt-header-detail">

              <h5>{stockInfo.name}</h5>
              <p className="p-stock">{stockInfo.sic_description}</p>

            </div>
            <div className="ticker">
              <h2>{stockInfo.ticker}</h2>
            </div>

          </div>

        </div>

        <div className="info-stocks">
          <p>check the info of the stocks </p>
        </div>

        <div className="info-container">

          <div className="wrapper">
            <div className="title-info">
              <p>City </p>
            </div>
            <div className="value">
              {stockInfo.address?.city ? (
                <p>{stockInfo.address.city}</p>
              ) : (
                <p>no data</p>
              )}

            </div>
          </div>

          <div className="wrapper">
            <div className="title-info">
              <p>State </p>
            </div>
            <div className="value">
              {stockInfo.address?.state ? (
                <p>{stockInfo.address.state}</p>
              ) : (
                <p>no data</p>
              )}

            </div>
          </div>

          <div className="wrapper">
            <div className="title-info">
              <p>Type of Market </p>
            </div>
            <div className="value">
              <p>{stockInfo.market}</p>
            </div>
          </div>
          <div className="wrapper">
            <div className="title-info">
              <p>total Employees</p>
            </div>
            <div className="value">
              <p>{stockInfo.total_employees}</p>
            </div>
          </div>

          <div className="wrapper">
            <div className="title-info">
              <p>primary exchange</p>
            </div>
            <div className="value">
              <p>{stockInfo.primary_exchange}</p>
            </div>
          </div>

          <div className="wrapper">
            <div className="title-info">
              <p>Market Cap</p>
            </div>
            <div className="value">
              <p>{stockInfo.market_cap}</p>
            </div>
          </div>

          <div className="wrapper">
            <div className="title-info">
              <p>local market</p>
            </div>
            <div className="value">
              <p>{stockInfo.locale}</p>
            </div>
          </div>

          <div className="wrapper">
            <div className="title-info">
              <p>list date</p>
            </div>
            <div className="value">
              <p>{stockInfo.list_date}</p>

            </div>
          </div>

          <div className="wrapper">

            <div className="title-info">
              <p>share class outstanding</p>
            </div>
            <div className="value">
              <p>{stockInfo.share_class_shares_outstanding}</p>

            </div>
          </div>

          <div className="wrapper">
            <div className="title-info">
              <p>Weighted shares outstanding</p>
            </div>
            <div className="value">
              <p>{stockInfo.weighted_shares_outstanding}</p>

            </div>
          </div>
        </div>
      </div>

    </>

  );
};

export default Details;
