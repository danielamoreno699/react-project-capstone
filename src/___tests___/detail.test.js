import * as React from 'react';
import { render, screen } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import Details from '../components/Details';
import { fetchStockInfo } from '../redux/stockSlice';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('../redux/stockSlice', () => ({
  fetchStockInfo: jest.fn(),
}));

describe('Details', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn());
    useSelector.mockReturnValue({
      stockInfo: {
        name: 'Stock Name',
        ticker: 'ABC',
      },
    });
  });

  afterEach(() => {
    useDispatch.mockClear();
    useSelector.mockClear();
    fetchStockInfo.mockClear();
  });

  test('renders stock details correctly', () => {
    render(
      <MemoryRouter initialEntries={['/stock/ABC']}>
        <Routes>
          <Route path="/stock/:stock" element={<Details />} />
        </Routes>
      </MemoryRouter>,
    );

    expect(fetchStockInfo).toHaveBeenCalledWith('ABC');
    expect(screen.getByText('Stock Name')).toBeInTheDocument();
    expect(screen.getByText('ABC')).toBeInTheDocument();
  });
});
