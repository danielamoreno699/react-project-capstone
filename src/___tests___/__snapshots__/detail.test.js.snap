// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details renders stock details correctly 1`] = `
<div
  class="detail-container"
  data-testid="details"
>
  <div
    class="back-symbol"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 11H6.414l5.293-5.293-1.414-1.414L2.586 12l7.707 7.707 1.414-1.414L6.414 13H21z"
      />
    </svg>
  </div>
  <div
    class="header-detail"
  >
    <div
      class="txt-container"
    >
      <div
        class="txt-header"
      >
        <h5>
          Stock Name
        </h5>
        <p
          class="p-stock"
        />
        <p
          class="url-txt"
        />
      </div>
      <div
        class="ticker"
      >
        <h2>
          ABC
        </h2>
      </div>
    </div>
  </div>
  <div
    class="info-stocks"
  >
    <p>
      check the info of the stocks 
    </p>
  </div>
  <div
    class="info-container"
  >
    <div
      class="wrapper"
    >
      <div
        class="title-info"
      >
        <p>
          City 
        </p>
      </div>
      <div
        class="value"
      >
        <p>
          no data
        </p>
      </div>
    </div>
    <div
      class="wrapper"
    >
      <div
        class="title-info"
      >
        <p>
          State 
        </p>
      </div>
      <div
        class="value"
      >
        <p>
          no data
        </p>
      </div>
    </div>
    <div
      class="wrapper"
    >
      <div
        class="title-info"
      >
        <p>
          Type of Market 
        </p>
      </div>
      <div
        class="value"
      >
        <p />
      </div>
    </div>
    <div
      class="wrapper"
    >
      <div
        class="title-info"
      >
        <p>
          total Employees
        </p>
      </div>
      <div
        class="value"
      >
        <p />
      </div>
    </div>
    <div
      class="wrapper"
    >
      <div
        class="title-info"
      >
        <p>
          primary exchange
        </p>
      </div>
      <div
        class="value"
      >
        <p />
      </div>
    </div>
    <div
      class="wrapper"
    >
      <div
        class="title-info"
      >
        <p>
          Market Cap
        </p>
      </div>
      <div
        class="value"
      >
        <p />
      </div>
    </div>
    <div
      class="wrapper"
    >
      <div
        class="title-info"
      >
        <p>
          local market
        </p>
      </div>
      <div
        class="value"
      >
        <p />
      </div>
    </div>
    <div
      class="wrapper"
    >
      <div
        class="title-info"
      >
        <p>
          list date
        </p>
      </div>
      <div
        class="value"
      >
        <p />
      </div>
    </div>
    <div
      class="wrapper"
    >
      <div
        class="title-info"
      >
        <p>
          share class outstanding
        </p>
      </div>
      <div
        class="value"
      >
        <p />
      </div>
    </div>
    <div
      class="wrapper"
    >
      <div
        class="title-info"
      >
        <p>
          Weighted shares outstanding
        </p>
      </div>
      <div
        class="value"
      >
        <p />
      </div>
    </div>
  </div>
</div>
`;
